### PR TITLE
Replace to-be deprecated popen4 with shellout

### PR DESCRIPTION
**Pull request by kaustubh-d**
Replace deprecated popen4 with shellout

### CODE REVIEW

**Comment by lamont-granquist** [comment]
@kaustubh-d can you take a shot at rebasing this off of master?

**Comment by lamont-granquist** [comment]
cc @chef/client-core @chef/client-engineers can you review the code as it is now?

File: lib/chef/provider/package/aix.rb [hunk 1]
```diff
@@ -87,19 +83,18 @@ def load_current_resource
...
+              fields = line.split(":")
+              @candidate_version = fields[2]
+              @new_resource.version(fields[2])
+              Chef::Log.debug("#{@new_resource} setting install candidate version to #{@candidate_version}")
             end
           end
-          unless status.exitstatus == 0
-            raise Chef::Exceptions::Package, "installp -L -d #{@new_resource.source} - #{status.inspect}!"
+          unless ret.exitstatus == 0
+            raise Chef::Exceptions::Package, "installp -L -d #{@new_resource.source} - #{ret.inspect}!"
```
Comment (danielsdeleo) [inline lib/chef/provider/package/aix.rb]: I don't have time to look this up just now, but what does `ShellOut#inspect` look like?

**Comment by danielsdeleo** [comment]
overall a pretty straightforward conversion, :+1: We should look at what `ShellOut#inspect` does, since that will be in the error messages. If it's something ugly, we may want to use `ShellOut#format_for_exception` instead. Side note, we could alternatively just use the `shell_out!` form, but that would change the exception type, unless we went all the way to something like:

```ruby
def run_some_pkg_command
  shell_out!("package command", returns: [0,1])
rescue ShellCommandFailed => e
  raise Chef::Exceptions::Package, e.message
end
```

But I think the improvement of just switching to shellout from popen4 is enough to merge without requiring the extra changes.

**Comment by tyler-ball** [comment]
Overall this looks pretty good :+1:

**Comment by lamont-granquist** [comment]
Marking this as ready-to-merge.  Still needs a rebase and bonus points to looking at Dan's comments, but its a Chef Software, Inc PR and we should get it across the finish line for 12.1.0 even if @kaustubh-d is AFK right now.

File: lib/chef/provider/package/aix.rb (hunk 1 above)
Comment (btm) [inline lib/chef/provider/package/aix.rb]:
```
irb(main):013:0> puts ret.inspect
<Mixlib::ShellOut#46614468: command: 'lslpp -lcq fooooo' process_status: #<Mixlib::ShellOut::Windows::ThingThatLooksSortOfLikeAProcessStatus:0x58a42b8 @exitstatus=1> stdout: '' stderr: ''lslpp' is not recognized as an internal or external command,
operable program or batch file.' child_pid: nil environment: {"LC_ALL"=>"C"} timeout: 600 user:  group:  working_dir:  >
=> nil
```

**Comment by kaustubh-d** [comment]
sure, working to rebase

File: lib/chef/provider/package/aix.rb (hunk 1 above)
Comment (tyler-ball) [inline lib/chef/provider/package/aix.rb]: Seems like we want `ret.command` and `ret.stderr` (and maybe `ret.stdout` for good measure)

Comment (danielsdeleo) [inline lib/chef/provider/package/aix.rb]: I think `#format_for_exception` will get us the info we want in a digestible format (same as shellout errors). https://github.com/chef/mixlib-shellout/blob/36c8f1c3d144a26f2f5f5f38a7ba59c10ef16a59/lib/mixlib/shellout.rb#L211-L223

**Comment by btm** [comment]
Rebased into #2924. It still looks a bit incomplete, like the replacement of Chef::Mixin::Command#handle_command_failures, and I think there's some places where shell_out needs to be shell_out!. I'm feeling like we keep working on that new branch for 12.2.0. Please post any comments on the new PR.

**Pull request closed without merging**